packages/apollo-engine-reporting/src/agent.ts

```diff
@@ -633,16 +677,45 @@ export class EngineReportingAgent<TContext = any> {
     });
 
     const statsReportKey = `# ${operationName || '-'}\n${signature}`;
+
     if (!report.tracesPerQuery.hasOwnProperty(statsReportKey)) {
       report.tracesPerQuery[statsReportKey] = new TracesAndStats();
       (report.tracesPerQuery[statsReportKey] as any).encodedTraces = [];
+      (report.tracesPerQuery[
+        statsReportKey
+      ] as any).statsWithContext = new StatsMap();
+    }
+
+    const traceCacheKey = {
+      statsReportKey,
+      statsBucket: DurationHistogram.durationToBucket(trace.durationNs),
+      endsAtMinute:
+        ((trace && trace.endTime && trace.endTime.seconds) || 0) % 60,
+    };
+
+    const convertTraceToStats =
+      reportData.traceCache.get(traceCacheKey) === true;
+
+    if (convertTraceToStats) {
+      (report.tracesPerQuery[statsReportKey] as any).statsWithContext.addTrace(
+        trace,
+      );
+    } else {
+      const protobufError = Trace.verify(trace);
+      if (protobufError) {
+        throw new Error(`Error encoding trace: ${protobufError}`);
+      }
+      const encodedTrace = Trace.encode(trace).finish();
+
+      // See comment on our override of Traces.encode inside of
+      // apollo-engine-reporting-protobuf to learn more about this strategy.
+      (report.tracesPerQuery[statsReportKey] as any).encodedTraces.push(
+        encodedTrace,
+      );
+      reportData.size +=
+        encodedTrace.length + Buffer.byteLength(statsReportKey);
+      reportData.traceCache.set(traceCacheKey, true);
```

Comment:
Is traceCache going to grow indefinitely?  That's not going to be good. Could do some sort of LRU thing. Or alternatively, instead of making trace.endTime.seconds be part of each entry's key, we can use just use the time that `addTrace` is called (presumably vaguely monotonic) and throw away the whole map every time we tick over to the next minute.

If we want to match the Apollo-cloud-service-side calculation which is based on the minute from the trace, and behave better when the clock turns back by a few seconds around a minute break, perhaps keep 2 or 3 maps of the most recent minutes?  This can just be a tiny map where every time we scan through it and drop sub-maps that are too old (it's `O(n)` but n=3 so whatever).

Also it's not *really* a "cache", it's more of a "seen" map